make rpc and dist-autograd multiprocess test to use both fork and spawn

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27136 make python udf serialization format to be binary plus tensor tables
* **#25656 make rpc and dist-autograd multiprocess test to use both fork and spawn**

spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

mrshenli commented on test/test_rpc.py:
```diff
@@ -1,8 +1,11 @@
 #!/usr/bin/env python3
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+import numpy
```
why do we need this import?

zhaojuanmao replied:
it tried to see whether it can fix test error about "numpy + intel MKL", but I removed in new diff, will see whether tests passed

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -4,29 +4,29 @@ namespace torch {
   }
 
   std::vector<char> generatePythonUDFResult(
     const Message& request) {
```
indent, does this fit into the line above?

zhaojuanmao replied:
ran clang-format in new diff

pritamdamania87 commented on test/common_distributed.py:
```diff
@@ -125,31 +125,52 @@ def setUpClass(cls):
-        self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
-        self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
+        if self.rank == self.MAIN_PROCESS_RANK:
```
Why do we need this change? Isn't the `setUp` method run only on the parent process?

zhaojuanmao replied:
this was added to check whether setup() is related after tests failures, it did not help, but I think it is no harm to keep there, if child process want to set up same variables as parent in the future, they can call setup() function

pritamdamania87 commented on test/common_distributed.py:
```diff
@@ -125,31 +125,52 @@ def setUpClass(cls):
+        # We're retrieving a corresponding test and executing it.
+        getattr(self, test_name)()
+        if multiprocessing.get_start_method() != "spawn":
+                sys.exit(0)
```
Doesn't this affect all c10d tests? Also, why are we checking for this after the test is done instead of before it?

zhaojuanmao replied:
c10d is not spawn, so it is not affected. we just change behavior for rpc, when it is spawn, we do not need to exit() here explicitly, as @xush6528 mentioned sys.exit(0) here is mainly for fork processes to avoid run teardown().

pritamdamania87 commented on test/test_rpc.py:
```diff
@@ -90,7 +92,17 @@ def wrapper(self):
     sys.version_info < (3, 0),
     "Pytorch distributed rpc package " "does not support python2",
 )
+@unittest.skipIf(C.is_asan, "Skip ASAN since torch + multiprocessing + fbcode doesn't work with asan")
```
don't mention fbcode in OSS code.

pritamdamania87 commented on test/test_rpc.py:
```diff
@@ -90,7 +92,17 @@ def wrapper(self):
+        super(RpcTest, cls). setUpClass()
+        try:
+            multiprocessing.set_start_method("spawn")
+        except Exception:
```
Why do we ignore this exception and continue? We should throw this exception and fail the test.

zhaojuanmao replied:
for some env, if there is no spawn, at least we should run fork?

xush6528 commented on test/common_distributed.py:
```diff
@@ -125,31 +125,52 @@ def setUpClass(cls):
         self.rank = rank
+        self.file_name = file_name
+
+        self.setUp()
```
Why do we need to call `self.setUp()`? What was missing?

zhaojuanmao commented:
although tests passed locally, CI still failed with two failures: 1) asan test, although I skipped asan  2) Numpy + Intel MKL incompatibility error

Tried to add some fast guessing changes (e.g., add setup() for child process as suspecting dependencies are not loaded properly; play with sys.exit(0) as suspecting process clean up has something wrong), but both of them did not work.

so will try following things:
1) download docker file to debug and try to fix for existing implementation, as it looks like it is close to finish
2) if it is found that it is hard to fix, will try to use torch.multiprocessing.spawn() - maybe it is better?

pietern commented:
@zhaojuanmao `torch.multiprocessing.spawn` is just a wrapper for doing `multiprocessing.Process` and joining them once done. It won't address more fundamental problems.

Btw -- on the previous version of this PR I left a comment regarding the runtime. Did you happen to check how much impact the spawn start method has on the total runtime of the distributed tests?

zhaojuanmao commented:
@pietern spawn is much slower, 28.67s vs 8.73s for test_rpc.

pietern commented:
@zhaojuanmao Thanks for checking. We should investigate if we can't switch to a mode where we launch processes once and reuse them across tests. There are downsides to doing that as well, but with such a runtime it's worth exploring in my opinion.

zhaojuanmao commented:
I tried to use torch.multiprocessing.spawn() in another PR for my investigation purpose, and all tests passed!

indeed, for torch.multiprocessing, both fork and spawn methods caught crash issue and run almost the same amount of time for test_rpc (28s). native multiprocessing fork run only 8s, but it did not catch the crash issue -- which means something is wrong. 

So I'm preferring to go ahead with torch.multiprocessing implementation in this diff. Is 28s acceptable?  I guess it should be fine now, spawn multi processes for each test case feel like it is more clean set up. The purpose of unit tests is to test correctness, so it should be fine to sacrifice some performance?

in this diff, make dist_autograd to use spawn as well. 

tried c10d tests, most of them works fine, a few ones are a little flaky some how, will address it in separate diff

mrshenli commented:
>  Is 28s acceptable?

It is OK to me for now, as `test_rpc.py` does not have too many tests and is unlikely to grow super large very soon, but let's create an issue to track this. Let's get @pietern's input on this as well.

mrshenli commented on test/multiprocessing_test_case.py:
```diff
@@ -0,0 +1,209 @@
+    return TIMEOUT_OVERRIDE.get(test_id.split(".")[-1], TIMEOUT_DEFAULT)
+
+
+class MultiProcessTestCase(TestCase):
```
Let's add some comments here to explain the difference between this `MultiProcessTestCase` and the one in `common_distributed.py`. It might be helpful to give them different names.

mrshenli commented on test/multiprocessing_test_case.py:
```diff
@@ -0,0 +1,209 @@
+}
+
+
+def skip_if_not_multigpu(func):
```
Can we import the one from `common_distributed`, or is this one different than that?

mrshenli commented on test/multiprocessing_test_case.py:
```diff
@@ -0,0 +1,209 @@
+    return wrapper
+
+
+def skip_if_lt_x_gpu(x):
```
ditto, and for all the functions below.

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -4,28 +4,28 @@ namespace torch {
-py::object module_;
-py::object runUDFFunction_;
-py::object loadResultFunction_;
+py::object runUDFFunction_ = py::none();
```
Is this intentional to put changes in `python_rpc_handler.cpp` into this PR?

zhaojuanmao replied:
yes, otherwise tests will fail

mrshenli reviewed (commented):
Do we have any new tests to make sure that `spawn` indeed catch crashes?

zhaojuanmao commented:
@mrshenli it is very hard to find another example to show this diff catch an issue that has not been caught previously right now, but we've already have one example, is not it? 

eventually common_distributed.py will be killed after test_c10d.py starts to use the new multiprocessing_test_case.py, so planning not to let them share same codes

zhaojuanmao commented:
@pietern regarding runtime concerns, you mentioned "We should investigate if we can't switch to a mode where we launch processes once and reuse them across tests. There are downsides to doing that as well" 

would you please elaborate more about your idea?  how could it be possible to insert each test into launched processes? you mentioned there are downsides, what kind of downsides in your mind?  

for unit tests, I guess runtime is not most concerned, correctness is more important?

mrshenli commented:
> eventually common_distributed.py will be killed after test_c10d.py starts to use the new multiprocessing_test_case.py

Just want to make sure I understand the plan here. By "will be killed" do you mean eventually, we will still have a file named `common_distributed.py`, but the `MultiProcessTestCase` will be replaced by the one implemented here? Otherwise, if the plan is to keep `MultiProcessTestCase` in a separate file named `multiprocessing_test_case.py`, we might still need a `common_distributed.py` file to keep utility functions and decorators.

zhaojuanmao commented:
@mrshenli multiprocessing_test_case.py has all the utilities and decorators that are same as common_distributed.py, it copies from common_distributed.py with different multiProcessTestCase class implementation. 

Do you prefer to still name "multiprocessing_test_case.py " as "common_distributed.py"?

mrshenli commented:
> Do you prefer to still name "multiprocessing_test_case.py " as "common_distributed.py"?

I feel things like `skip_if_lt_x_gpu` and `requires_gloo` should be independent concepts of `multiprocessing_test_case`. "common_distributed.py" sounds like a more proper place to hold those utility functions.

zhaojuanmao commented:
> > Do you prefer to still name "multiprocessing_test_case.py " as "common_distributed.py"?
> 
> I feel things like `skip_if_lt_x_gpu` and `requires_gloo` should be independent concepts of `multiprocessing_test_case`. "common_distributed.py" sounds like a more proper place to hold those utility functions.

Sounds good, in next PR I will still name it as "common_distributed.py" once test_10d.py uses spawn as well

mrshenli commented:
@pytorchbot retest this please

mrshenli reviewed (approved):
LGTM. Thanks for adding this! 

It seems some CI tests are not configured properly, not sure if a rebase would help. Please make sure all CI tests pass before landing.

pietern commented on test/common_distributed.py:
```diff
@@ -129,26 +127,48 @@ def setUp(self):
+                name='process ' + str(rank),
+                args=(rank, self._current_test_name(), self.file_name))
+            process.start()
+            self.processes.append(process)
```
These two do the same, but with a different context (default vs custom). Please consolidate into 1 base function.

Could you add some comment about `torch.multiprocessing.get_context` only being available on Python 3 (also see https://docs.python.org/3/library/multiprocessing.html#multiprocessing.get_context).

If I recall correctly, the Python 3 thing is why you use spawn in the RPC tests and fork in the c10d tests, right?

zhaojuanmao replied:
1. I tried to consolidate them  by passing "ctx = get_context("fork"/"spawn")" to base function, but it failed in python2 tests, get_context() is not available in python2. 

so still separating it into two: 
for fork, call torch.multiprocessing.process(), in default it will use fork in python2 and python3

for spawn, call ctx = get_context("fork"/"spawn"), then ctx.process()

2. we can still make c10d to use spawn in python 3 test and use fork in python 2 test in following up diff

zhaojuanmao replied:
also, the reason I did not use native multiprocessing.process() with start_method(), because it will have two weird test failures I mentioned before: 1) asan failure even I skipped asan tests 2) MKL incompatiable with numpy library.

pietern replied:
I was thinking something like this. It uses the `Process` class on the module for Python 2. That should work right?

```python
def _start_processes(self, proc):
  self.processes = []
  for rank in range(int(self.world_size)):
    process = proc(
      target=self.__class__._run,
      name='process ' + str(rank),
      args=(rank, self._current_test_name(), self.file_name))
    process.start()
    self.processes.append(process)

def _fork_processes(self):
  if six.PY3:
    proc = torch.multiprocessing.get_context("fork").Process
  else:
    proc = torch.multiprocessing.Process
  self._start_processes(proc)

def _spawn_processes(self):
  if six.PY3:
    proc = torch.multiprocessing.get_context("spawn").Process
  else:
    raise RuntimeError("Cannot use spawn start method with Python 2")
  self._start_processes(proc)
```

pietern replied:
I think those failures indicate something else is wrong. Did you end up filing issues for them?

zhaojuanmao replied:
awesome, it works! just updated the PR, still did not get used to pass callable constructor into functions in python

zhaojuanmao replied:
@pietern  I guess it is related to test setup on OSS? because these tests are not even ran yet, it failed at setupClass, I guess.  To file the issue, I will need to write a separate simple program and reproduce it in CI tests

zhaojuanmao replied:
right now, I have 3 pending issues to file :), I feel it is better to write some small reproducible sample codes to file, they are still in my TODO list:

1. call sparse tensor coalesce, indices tensor is not contiguous;  
2. multiprocess test with start_method() failed in CI tests; 
3. caffe2.python._import_c_extension can not be imported in pytorch_macos_10_13_py3_test

pietern commented on test/common_distributed.py:
```diff
@@ -129,26 +127,48 @@ def setUp(self):
+            process.start()
+            self.processes.append(process)
+
+    def _spawn_process(self):
```
Can you rename these to be the plural version? If you read the constructor code in the subclasses, it reads as if they're spawning a single process, but in fact they're spawning multiple.

zhaojuanmao commented:
@pytorchbot retest this please

zhaojuanmao commented:
@pytorchbot retest this please

zhaojuanmao commented:
@pytorchbot retest this please

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

pritamdamania87 commented on test/common_distributed.py:
```diff
@@ -129,26 +128,51 @@ def setUp(self):
+
+    def _fork_processes(self):
+        if six.PY3:
+            proc = torch.multiprocessing.get_context("fork").Process
```
nit: call this `mp_context` instead of `proc`

pietern replied:
It's not the context though, it's the `Process` class/function on that context.

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot retest this please

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

mrshenli commented:
```
02:34:22 Running test_rpc_fork ... [2019-10-01 02:34:23.013131]
02:34:23 Traceback (most recent call last):
02:34:23   File "test_rpc_fork.py", line 4, in <module>
02:34:23     from rpc_test import RpcTest
02:34:23   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\rpc_test.py", line 11, in <module>
02:34:23     from torch.distributed.rpc import RpcBackend
02:34:23   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\distributed\rpc.py", line 3, in <module>
02:34:23     from . import invoke_rpc_builtin, invoke_rpc_python_udf, invoke_remote_builtin
02:34:23 ImportError: cannot import name 'invoke_rpc_builtin'
02:34:23 Traceback (most recent call last):
02:34:23   File "run_test.py", line 442, in <module>
02:34:23     main()
02:34:23   File "run_test.py", line 434, in main
02:34:23     raise RuntimeError(message)
02:34:23 RuntimeError: test_rpc_fork failed!
```

I think the above test failure is because we don't build rpc code for windows (see [this line](https://github.com/pytorch/pytorch/blob/master/torch/CMakeLists.txt#L225)). Let's skip RPC tests for windows too.

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot retest this please

zhaojuanmao commented:
landing the  diff, the failure is not related

zhaojuanmao committed:
Update on "make rpc and dist-autograd multiprocess test to use both fork and spawn"


spawn multiprocessing can catch some issues that fork multiprocessing can not catch, meanwhile fork can work properly with asan tests, but spawn multiprocessing can not work with asan tests for some use cases right now. 

so this diff adding support to launch both spawn and fork tests in multiProcessingTestCase class, also let test_rpc and test_dist_autograd to run both spawn and fork tests

Differential Revision: [D17086007](https://our.internmc.facebook.com/intern/diff/D17086007/)

[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot retest this please

zrphercule commented:
Hi @zhaojuanmao it seems this pr breaks pytorch master ci test, wonder if you are fixing it, or should I revert it?

mrshenli commented:
@zrphercule we are landing #27158 to disable the flaky test now. Please don't revert.

mrshenli commented:
#27158 is landed
